src/main/host/network_interface.c

```diff
@@ -94,135 +80,32 @@ static void _compatsocket_unrefTaggedVoid(void* taggedSocketPtr) {
-    /* We have room to add more tokens. */
-    bucket->bytesRemaining += bucket->bytesRefill;
-    /* Make sure we stay within capacity. */
-    if(bucket->bytesRemaining > bucket->bytesCapacity) {
-        bucket->bytesRemaining = bucket->bytesCapacity;
-    }
-}
-
-static void
-_networkinterface_consumeTokenBucket(NetworkInterfaceTokenBucket* bucket,
-                                     guint64 bytesConsumed) {
-    if (bytesConsumed >= bucket->bytesRemaining) {
-        bucket->bytesRemaining = 0;
-    } else {
-        bucket->bytesRemaining -= bytesConsumed;
-    }
-}
-
-static void _networkinterface_scheduleRefillTask(NetworkInterface* interface, Host* host,
-                                                 TaskCallbackFunc func, SimulationTime delay) {
-    TaskRef* refillTask = taskref_new_bound(host_getID(host), func, interface, NULL, NULL, NULL);
-    worker_scheduleTaskWithDelay(refillTask, host, delay);
-    taskref_drop(refillTask);
-    interface->isRefillPending = TRUE;
-}
-
-static void _networkinterface_scheduleNextRefill(NetworkInterface* interface, Host* host) {
-    SimulationTime now = worker_getCurrentSimulationTime();
-    SimulationTime interval = _networkinterface_getRefillInterval();
-
-    /* This computes the time that the next refill should have occurred if we
-     * were to always call the refill function every interval, even though in
-     * practice we stop scheduling refill tasks once we notice we are idle. */
-    SimulationTime offset = now - interface->timeStartedRefillingBuckets;
-    SimulationTime relTimeSincelastRefill = offset % interval;
-    SimulationTime relTimeUntilNextRefill = interval - relTimeSincelastRefill;
-
-    /* call back when we need the next refill */
-    _networkinterface_scheduleRefillTask(
-        interface, host, _networkinterface_refillTokenBucketsCB, relTimeUntilNextRefill);
-}
-
-static gboolean _networkinterface_isRefillNeeded(NetworkInterface* interface) {
-    gboolean sendNeedsRefill = interface->sendBucket.bytesRemaining <
-                               interface->sendBucket.bytesCapacity;
-    gboolean receiveNeedsRefill = interface->receiveBucket.bytesRemaining <
-                                  interface->receiveBucket.bytesCapacity;
-    return (sendNeedsRefill || receiveNeedsRefill) &&
-           !interface->isRefillPending;
-}
-
-static void _networkinterface_scheduleNextRefillIfNeeded(NetworkInterface* interface, Host* host) {
-    if (_networkinterface_isRefillNeeded(interface)) {
-        _networkinterface_scheduleNextRefill(interface, host);
-    }
+static TokenBucket* _networkinterface_create_tb(uint64_t bwKiBps) {
+    uint64_t refill_interval_nanos = SIMTIME_ONE_MILLISECOND;
+    uint64_t refill_size = bwKiBps * 1024 / 1000;
+    // CONFIG_MTU ensures we don't "waste" any partial packet byte-chunks we
+    // had left from last round when we do the refill.
+    uint64_t capacity = refill_size + CONFIG_MTU;
```

Comment:
Can you add an example here? I'm not clear on what the `CONFIG_MTU` helps with here. Could this also be solved by supporting partial refills (for example if only `refill_interval/2` time has taken place, to add `refill_size/2` balance)?

Reply:
The `CONFIG_MTU` part represents a "burst allowance" which is common in token buckets. Only the `capacity` of the bucket is increased by `CONFIG_MTU`, not the `refill_size`. Therefore, the long term rate limit enforced by the token bucket (configured by `refill_size`) is not affected much.

What the burst allowance ensures is that we don't lose tokens that are unused because we don't fragment packets. If we set the capacity of the bucket to exactly the refill size (i.e., without the `CONFIG_MTU` burst allowance) and there are only 1499 tokens left in this sending round, a full packet would not fit. The next time the bucket refills, it adds `refill_size` tokens but in doing so 1499 tokens would fall over the top of the bucket; these tokens would represent wasted bandwidth, and could potentially accumulate in every refill interval leading to a significantly lower achievable bandwidth.

A downside of the `CONFIG_MTU` burst allowance is that the sending rate could possibly become "bursty" with a behavior such as:
- interval 1: send `refill_size` + `CONFIG_MTU` bytes, sending over the allowance by 1500 bytes
- refill: `refill_size` token gets added to the bucket
- interval 2: send `refill_size` - `CONFIG_MTU` bytes, sending under the allowance by 1500 bytes
- refill: `refill_size` token gets added to the bucket
- interval 3: send `refill_size` + `CONFIG_MTU` bytes, sending over the allowance by 1500 bytes
- repeat

So it could become less smooth and more "bursty" even though the long term average is maintained. But I don't think this would happen much in practice, and we are already batching sends for performance reasons anyway.

Note:
- Partial refills are essentially just setting a different batch time, and lower batch times means more refills and worse simulator performance.
- I am not changing the behavior in the PR, just carrying over the implementation from the C network interface code.

Reply:
Okay, this makes a lot of sense now!